### PR TITLE
[IGA] 3p Shell stress recovery 

### PR DESCRIPTION
**Pull request by rickyaristio**
**Description**
In this PR, I added some functions related to stress recovery to iga shell 3p element based on @MichaelLoibl prior implementation. 

**Changelog**
- Added CalculateOnIntegrationPoints to Shell_3p_element.
- Added new variables (PK2_STRESS, CAUCHY_STRESS, SHEAR_FORCE, etc).
- Modify the Scordelis roof test.

### CODE REVIEW

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 1]
```diff
@@ -753,6 +1371,26 @@ namespace Kratos
         }
     }
 
+    void Shell3pElement::CalculateSecondDerivativesOfBaseVectors(
+        const Matrix& rDDDN_DDDe,
+        array_1d<double, 3>& rDDa1_DD11,
+        array_1d<double, 3>& rDDa1_DD12,
+        array_1d<double, 3>& rDDa2_DD21,
+        array_1d<double, 3>& rDDa2_DD22)
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp]:
```suggestion
        array_1d<double, 3>& rDDa2_DD22) const
```

File: applications/IgaApplication/custom_python/iga_python_application.cpp [hunk 2]
```diff
@@ -46,6 +46,16 @@ PYBIND11_MODULE(KratosIgaApplication, m)
...
+    KRATOS_REGISTER_IN_PYTHON_SYMMETRIC_2D_TENSOR_VARIABLE_WITH_COMPONENTS(m, PK2_STRESS)
+    KRATOS_REGISTER_IN_PYTHON_SYMMETRIC_2D_TENSOR_VARIABLE_WITH_COMPONENTS(m, CAUCHY_STRESS)
+    KRATOS_REGISTER_IN_PYTHON_SYMMETRIC_2D_TENSOR_VARIABLE_WITH_COMPONENTS(m, CAUCHY_STRESS_TOP)
+    KRATOS_REGISTER_IN_PYTHON_SYMMETRIC_2D_TENSOR_VARIABLE_WITH_COMPONENTS(m, CAUCHY_STRESS_BOTTOM)
+    KRATOS_REGISTER_IN_PYTHON_SYMMETRIC_2D_TENSOR_VARIABLE_WITH_COMPONENTS(m, MEMBRANE_FORCE)
+    KRATOS_REGISTER_IN_PYTHON_SYMMETRIC_2D_TENSOR_VARIABLE_WITH_COMPONENTS(m, INTERNAL_MOMENT)
+
+    KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, SHEAR_FORCE_1)
+    KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, SHEAR_FORCE_2)
+
```
Comment (philbucher) [inline applications/IgaApplication/custom_python/iga_python_application.cpp:58]: please make sure this is consistent with StructuralMechanics
=> Check the variable-name used here: https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 3]
```diff
@@ -590,6 +773,441 @@ namespace Kratos
...
+            / pow(m_dA_vector[IntegrationPointIndex], 2);
+        rDCurvature_D1[0] = inner_prod(DDa1_DD11, a3) + inner_prod(Da1_D1, Da3_D1);
+        rDCurvature_D1[1] = inner_prod(DDa2_DD21, a3) + inner_prod(Da2_D2, Da3_D1);
+        rDCurvature_D1[2] = inner_prod(DDa1_DD12, a3) + inner_prod(Da1_D2, Da3_D1);
+        rDCurvature_D2[0] = inner_prod(DDa1_DD12, a3) + inner_prod(Da1_D1, Da3_D2);
+        rDCurvature_D2[1] = inner_prod(DDa2_DD22, a3) + inner_prod(Da2_D2, Da3_D2);
+        rDCurvature_D2[2] = inner_prod(DDa2_DD21, a3) + inner_prod(Da1_D2, Da3_D2);
+    }
+    
+    void Shell3pElement::CalculateDerivativeOfCurvatureActual(
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp:1124]: this is a very long fct.
Can you refactor a bit?

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 4]
```diff
@@ -590,6 +773,441 @@ namespace Kratos
...
+        IndexType IntegrationPointIndex,
+        array_1d<double, 3>& rDCurvature_D1,
+        array_1d<double, 3>& rDCurvature_D2,
+        const Matrix& rHessian,
+        const KinematicVariables& rKinematicVariables)
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp]:
```suggestion
        const KinematicVariables& rKinematicVariables) const
```

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 5]
```diff
@@ -590,6 +773,441 @@ namespace Kratos
...
+
+        array_1d<double, 3> bending_stress_pk2_cov = prod(T_car_to_cov, rPK2BendingStressCartesian);
+        array_1d<double, 3> bending_stress_cau_cov = bending_stress_pk2_cov / detF;
+        array_1d<double, 3> bending_stress_cau_car = prod(T_cov_to_car, bending_stress_cau_cov);
+
+        rCauchyMembraneStressesCartesian = membrane_stress_cau_car;
+        rCauchyBendingStressesCartesian = bending_stress_cau_car;
+    }
+
+    void Shell3pElement::CalculateShearForce(
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp:914]: same

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 6]
```diff
@@ -590,6 +773,441 @@ namespace Kratos
...
+        IndexType IntegrationPointIndex,
+        array_1d<double, 2>& rq, 
+        const KinematicVariables& rKinematicVariables,
+        const ConstitutiveVariables& rConstitutiveVariablesCurvature)
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp]:
```suggestion
        const ConstitutiveVariables& rConstitutiveVariablesCurvature) const
```

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 7]
```diff
@@ -590,6 +773,441 @@ namespace Kratos
...
+    ///@name Stress recovery
+    ///@{
+    
+    void Shell3pElement::CalculateCauchyStress(
+        IndexType IntegrationPointIndex,
+        array_1d<double, 3>& rCauchyMembraneStressesCartesian, 
+        array_1d<double, 3>& rCauchyBendingStressesCartesian, 
+        const array_1d<double, 3>& rPK2MembraneStressCartesian,
+        const array_1d<double, 3>& rPK2BendingStressCartesian,
+        const KinematicVariables& rKinematicVariables)
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp]:
```suggestion
        const KinematicVariables& rKinematicVariables) const
```

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 8]
```diff
@@ -590,6 +773,441 @@ namespace Kratos
         }
     }
 
+    ///@}
+    ///@name Stress recovery
+    ///@{
+    
+    void Shell3pElement::CalculateCauchyStress(
+        IndexType IntegrationPointIndex,
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp]:
```suggestion
        const IndexType IntegrationPointIndex,
```

File: applications/IgaApplication/custom_elements/shell_3p_element.cpp [hunk 9]
```diff
@@ -293,6 +433,49 @@ namespace Kratos
         rT(2, 2) = 2 * (G(0, 0) * G(1, 1) + G(0, 1) * G(1, 0));
     }
 
+    /* Computes the transformation matrix T from the local cartesian basis to
+    *  the local cartesian basis.
+    */
+    void Shell3pElement::CalculateTransformationFromCovariantToCartesian(
+        const KinematicVariables& rKinematicVariables,
+        Matrix& rTCovToCar
+    )
```
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_3p_element.cpp]:
```suggestion
    ) const
```

**Review by philbucher — COMMENTED** [review]
looks ok to me, but please make the fcts `const` if possible and check that the variables are consistent with StructuralMechanics

**Comment by philbucher** [comment]
codacy also has suggestions

**Comment by rickyaristio** [comment]
> looks ok to me, but please make the fcts `const` if possible and check that the variables are consistent with StructuralMechanics

**Comment by rickyaristio** [comment]
I apologize for being inactive for this long. I have pushed major changes related to stress recovery and added new examples to checking the moment and shear forces.

**Comment by rickyaristio** [comment]
> looks ok to me, but please make the fcts `const` if possible and check that the variables are consistent with StructuralMechanics

I have already addressed the issues. Thanks.

**Review by tteschemacher — APPROVED** [review]
Thank you @rickyaristio for this really great work!

**Pull request merged**